secp256kfun/src/poly.rs

```diff
@@ -0,0 +1,260 @@
+    s, Point, Scalar, G,
+};
+
+/// Create a vector of points by multiplying each scalar by `G`.
+///
+/// # Example
+///
+/// ```
+/// use secp256kfun::{g, poly, s, Scalar, G};
+/// let secret_poly = (0..5)
+///     .map(|_| Scalar::random(&mut rand::thread_rng()))
+///     .collect::<Vec<_>>();
+/// let point_poly = poly::to_point_poly(&secret_poly);
+/// ```
+pub fn to_point_poly(scalar_poly: &[Scalar]) -> Vec<Point> {
+    scalar_poly.iter().map(|a| g!(a * G).normalize()).collect()
+}
+
+/// Generate a [`Scalar`] polynomial for key generation
+///
+/// [`Scalar`]: crate::Scalar
+pub fn generate_scalar_poly(threshold: usize, rng: &mut impl RngCore) -> Vec<Scalar> {
+    (0..threshold).map(|_| Scalar::random(rng)).collect()
+}
+
+/// Evaluate a scalar polynomial defined by coefficients, at some scalar index.
+///
+/// The polynomial coefficients begin with the smallest degree term first (the constant).
+pub fn scalar_poly_eval(
+    poly: &[Scalar],
+    x: Scalar<impl Secrecy, impl ZeroChoice>,
+) -> Scalar<Secret, Zero> {
+    poly.iter()
+        .fold((s!(0), s!(1).mark_zero()), |(eval, xpow), coeff| {
+            (s!(eval + xpow * coeff), s!(xpow * x))
+        })
+        .0
+}
+
+/// Evaluate a point polynomial defined by coefficients, at some scalar index.
+///
+/// The polynomial coefficients begin with the smallest degree term first (the constant).
+pub fn point_poly_eval(
+    poly: &[Point<impl PointType, Public, impl ZeroChoice>],
+    x: Scalar<Public, impl ZeroChoice>,
+) -> Point<NonNormal, Public, Zero> {
+    let xpows = core::iter::successors(Some(s!(1).public().mark_zero()), |xpow| {
+        Some(s!(xpow * x).public())
+    })
+    .take(poly.len())
+    .collect::<Vec<_>>();
+    crate::op::lincomb(&xpows, poly.iter())
+}
+
+/// Calculate the lagrange coefficient for participant with index x_j and other signers indexes x_ms
+pub fn lagrange_lambda(
+    x_j: Scalar<impl Secrecy>,
+    x_ms: impl Iterator<Item = Scalar<impl Secrecy>>,
+) -> Scalar<Public> {
+    x_ms.fold(Scalar::one(), |acc, x_m| {
+        let denominator = s!(x_m - x_j)
+            .non_zero()
+            .expect("indexes must be unique")
+            .invert();
+        s!(acc * x_m * denominator).public()
+    })
+}
+
+/// Find the coefficients of the polynomial that interpolates a set of points at unique indexes.
+///
+/// Panics if the indexes are not unique.
+///
+/// A vector with a tail of zero coefficients means the interpolation was overdetermined.
+pub fn interpolate_point_polynomial(
+    indexes: Vec<Scalar<impl Secrecy, impl ZeroChoice>>,
+    points: Vec<Point>,
+) -> Vec<Point<impl PointType, Public, Zero>> {
+    // Get each lagrange basis polynomial from the product of coefficients:
+    //      l_j(x) = Product[ (x-x_m)/(x_j-x_m), j!=m]
+    // Or
+    //      l_j(x) = Product[ a_m*x + b_m, j!=m], where a_m = 1/(x_j-x_m) and b_m = -x_m*a_m.
+    let basis_polynomials: Vec<_> = indexes
```

Comment:
I think this would be easier to understand as a single `for` loop that starts with `sum` an identity poly (all Point::zero) and figures out the basis poly in each iteration and adds the basis poly to `sum`.

Reply:
Do you mean sum the basis polynomials as we go? I think this would be more confusing, I like how it is broken out into steps: find unit basis polys, multiply to intersect points, then sum. Or do you mean something else?

Reply:
> I like how it is broken out into steps: find unit basis polys, multiply to intersect points, then sum

Summing the scaled basis polynomials as we go is also broken up into steps. You figure out the basis poly, scale by the y coordinate and then add it back to the sum. Each step seems simpler also because it doesn't involve an iteration. I think implementing an algorithm in the fewest number of iterations usually leads to the simplest to understand code and maps to the textbook reading of the algorithm (there is only one loop):

![image](https://github.com/LLFourn/secp256kfun/assets/3734083/95c9f1fd-1a44-459e-a53a-2257a5803000)

This is just a point of style and not a huge concern for me so just roll with what you feel. I understand that every rust programmer has to go through the use successive iterator combinators everywhere stage of their career!

Reply:
Ah got you, changed. Better?